File: 30-Day LeetCoding Challenge/Week 3 April 15th–April 21st /Valid Parenthesis String.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        suru = shes = 0
        l = len(s)
        for i in range(0,l):
            if s[i] == '(' or s[i] == '*':
                suru += 1
            else:
                suru -= 1
            if s[l-1-i] == ')' or s[l-1-i] == '*':
                shes += 1
            else:
                shes -= 1
            if suru<0 or shes<0:
                return False
        return True
            
          
            
            
        
        

File: 30-Day LeetCoding Challenge/Week 3 April 15th–April 21st /Valid Parenthesis String.cpp
class Solution {
public:
    bool checkValidString(string s) {
        
        stack<int> opn, ast;
        for (int i=0; i<s.length();i++)
        {
            if (s[i] == ')')
            {
                if(!opn.empty()){
                    opn.pop();
                }
                
                else if(!ast.empty()){
                    ast.pop();
                }
                else
                    return false;
            }
            else if(s[i]=='(')
                opn.push(i);
            else
                ast.push(i);
        }
        
        while( !opn.empty() and !ast.empty()){
            if(opn.top() > ast.top())
                return false;
            opn.pop();
            ast.pop();
        }
        
        return opn.empty();
    }
};